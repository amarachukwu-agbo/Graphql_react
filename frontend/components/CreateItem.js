import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import Router from 'next/router';

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;
export default class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0
  }

  handleChange = (e) => {
    const { name, value, type } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({
      [name]: val
    });
  }

  uploadImage = async (e) => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');
    const res = await fetch('https://api.cloudinary.com/v1_1/ama-hello-books-v2/image/upload', {
      method: 'POST',
      mode: 'cors',
      body: data
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    });
  }
  render() {
    const { title, price, description, image } = this.state;
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
      {(createItem, { loading, error}) => (
        <Form onSubmit={async (e) => {
          e.preventDefault();
          const res = await createItem();
          Router.push({
            pathname: '/item',
            query: { id: res.data.createItem.id}
          });
        }}>
          <h2> Sell an item</h2>
          <Error error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
          <label htmlFor="file">
            File
            </label>
            <input
              type="file"
              id="file"
              name="file"
              placeholder="Upload an image"
              required
              onChange={this.uploadImage}
            />
            {
              image && <img src={image} alt="uploaded image" width="50px"/>
            }
            <label htmlFor="title">
            Title
            </label>
            <input
              type="text"
              id="title"
              name="title"
              placeholder="Title"
              required
              value={title}
              onChange={this.handleChange}
            />

            <label htmlFor="price">
            Price
            </label>
            <input
              type="number"
              id="price"
              name="price"
              placeholder="Price"
              required
              value={price}
              onChange={this.handleChange}
            />

            <label htmlFor="description">
            Description
            </label>
            <textarea
              type="text"
              id="description"
              name="description"
              placeholder="Enter a description"
              required
              value={description}
              onChange={this.handleChange}
            />
            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      )}
      </Mutation>
    )
  }
}
