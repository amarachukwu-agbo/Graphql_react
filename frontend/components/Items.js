import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Item from './Item';
import Pagination from './Pagination';
import { perPage } from '../config';


export const ITEMS_QUERY = gql`
  query ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}){
    items(skip: $skip, first: $first) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  margin: 0 auto;
  max-width: ${props => props.theme.maxWidth};
`;

export default class Items extends Component {
  render() {
    const { page } = this.props;
    return (
      <Center>
      <Query query={ITEMS_QUERY} variables={{
        skip: (page - 1) * perPage
      }}>
        {({ error, loading, data }) => {
          if (loading) return <p>Loading... </p>
          if (error) return <p>{`Error: ${error.message}`}</p>
          console.log(data, 'data =====');
          return data.items.length ? (
            <div>
              <Pagination page={page}/>
              <ItemsList>
                {
                  data.items.map(item => <Item key ={item.id} item={item}/>)
                }
              </ItemsList>
              <Pagination page={page}/>
            </div>
          ) : <p>No items have been created :(</p>
        }}
      </Query>
      </Center>
    )
  }
}
