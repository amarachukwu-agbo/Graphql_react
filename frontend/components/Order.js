import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import { format } from 'date-fns';
import gql from 'graphql-tag';
import Head from 'next/head';
import OrderStyles from './styles/OrderStyles';
import Error from '../components/ErrorMessage';
import formatMoney from '../lib/formatMoney';

export const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      total
      charge
      items {
        id
        quantity
        price
        title
        image
      }
    }
  }
`;


export default class Order extends Component {
  static propTypes = {
    orderId: PropTypes.string.isRequired
  }
  render() {
    const { orderId } = this.props;
    return (
      <Query query={SINGLE_ORDER_QUERY} variables={{id: orderId}}>
        {({ data: {order}, loading, error }) => {
          if (error) return <Error error ={error} />
          if (loading) return <p>Loading ....</p>
          return (
            <OrderStyles>
              <Head>
                <title>Sick Fits - Order {orderId}</title>
              </Head>
              <p>
                <span>Order ID: </span>
                <span>{ orderId }</span>
              </p>
              <p>
                <span>Charge: </span>
                <span>{ order.charge }</span>
              </p>
              <p>
                <span>Date: </span>
                <span>{ format(order.createdAt, 'MMMM d, YYYY h:mm a') }</span>
              </p>
              <p>
                <span>Order Total: </span>
                <span>{ formatMoney(order.total) }</span>
              </p>
              <p>
                <span>Item Count: </span>
                <span>{ order.items.length }</span>
              </p>
              <div className="items">
                {
                  order.items.map(item => (
                    <div className="order-item" key={item.id}>
                      <img src={item.image} alt={item.title}/>
                      <div className="item-details">
                        <h2>{item.title}</h2>
                        <p>Quantity: {item.quantity}</p>
                        <p>Each: {formatMoney(item.price)}</p>
                        <p>Subtotal: {formatMoney(item.quantity * item.price)}</p>
                      </div>
                    </div>
                  ))
                }
              </div>
            </OrderStyles>
          )
        }}
      </Query>
    )
  }
}
