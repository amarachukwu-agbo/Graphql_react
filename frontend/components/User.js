import React from 'react'
import PropTypes from 'prop-types'
import gql from 'graphql-tag';
import { Query } from 'react-apollo';

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      orders
      cart {
        id
        quantity
        item {
          id
          price
          description
          image
          title
        }
      }
    }
  }

`;
const User = props => (
  <Query query={CURRENT_USER_QUERY}>
    {(payload) => {
      return props.children(payload)
    }}
  </Query>
);

User.propTypes = {
  children: PropTypes.func.isRequired
}

export default User;

