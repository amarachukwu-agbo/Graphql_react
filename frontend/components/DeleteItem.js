import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
      title
    }
  }
`;

export default class DeleteItem extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired
  }
  update(cache, payload) {
    const { items } = cache.readQuery({ query: ITEMS_QUERY });
    const requiredItems = items
      .filter(item => item.id !== payload.data.deleteItem.id)
    cache.writeQuery({ query: ITEMS_QUERY, data: { items: requiredItems } })
  }
  render() {
    const { id, children } = this.props;
    return (
      <Mutation mutation={DELETE_ITEM_MUTATION}
        variables={{id}}
        update={this.update}
        >
      {(deleteItem) => {
        return (
          <button onClick={() => {
            if (confirm('Are you sure you want to delete this item')) {
              deleteItem().catch(err => alert(err));
            }
          }}>
            { children }
          </button>
        )
      }}
      </Mutation>
    )
  }
}
