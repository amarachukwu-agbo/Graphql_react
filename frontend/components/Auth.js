import React from 'react';
import { Query } from 'react-apollo';
import { CURRENT_USER_QUERY } from './User';
import SignIn from '../components/SignIn';

const Auth = ({ children }) => {
  return (
    <Query query={CURRENT_USER_QUERY}>
      {({ data , loading}) => {
        if (loading) return <p>Loading ...</p>
        if (data && !data.me) return <SignIn />
        return children;
      }}
    </Query>
  );
};

export default Auth;