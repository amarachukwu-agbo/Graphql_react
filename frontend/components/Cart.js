import React, { Component } from 'react';
import gql from 'graphql-tag';
import StripeCheckout from 'react-stripe-checkout';
import Router from 'next/router';
import NProgress from 'nprogress';
import { Query, Mutation } from 'react-apollo';
import CartStyles from './styles/CartStyles';
import SickButton from './styles/SickButton';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';
import User, { CURRENT_USER_QUERY } from './User';
import CartItem from './CartItem';
import calculateTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

export const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`;

export const CREATE_ORDER = gql`
  mutation CREATE_ORDER($token: String!) {
    createOrder(token: $token) {
      id
      items {
        image
        description
        title
      }
      total
      charge
    }
  }
`;

const totalItems = (cart) => {
  return cart.reduce((tally, item) => tally + item.quantity, 0)
}
class Cart extends Component {
  onToken = async (token, createOrder) => {
    NProgress.start();
    const order = await createOrder({
      variables: {
        token: token.id
      }
    }).catch(error => alert(`Error: ${error}`));
    Router.push({
      pathname: '/order',
      query: { id: order.data.createOrder.id }
    });
  }
  render() {
    return (
      <User>
        {({ data } ) => {
          const { me } = data;
          if (!me) return null;
          return(
            <Mutation mutation={TOGGLE_CART_MUTATION}>
              {(toggleCart) => (
              <Query query={LOCAL_STATE_QUERY}>
                {({ data }) => (
                  <CartStyles open={data.cartOpen}>
                    <header>
                      <CloseButton title="close" onClick={toggleCart}>&times;</CloseButton>
                      <Supreme>{me.name}'s Cart</Supreme>
                      <p>You have {me.cart.length} item{me.cart.length === 1 ? '': 's'} in your cart</p>
                    </header>
                    <ul>
                      { me.cart.map(item => <CartItem key={item.id} cartItem={item}/>) }
                    </ul>
                    <footer>
                      {
                        me.cart.length ? (
                          <>
                            <p>{formatMoney(calculateTotalPrice(me.cart))}</p>
                            <Mutation mutation={CREATE_ORDER} refetchQueries={[{query: CURRENT_USER_QUERY }]}>
                              {(createOrder) => (
                                <StripeCheckout
                                  name="Sick Fits"
                                  description={`Order of ${totalItems(me.cart)} items`}
                                  token={(stripeToken) => this.onToken(stripeToken, createOrder)}
                                  stripeKey="pk_test_4ApAJihRk5yShkglGcJ0ZSXV00oEwKYLu4"
                                  amount={calculateTotalPrice(me.cart)}
                                  image={me.cart.length && me.cart[0].item && me.cart[0].item.image}
                                  currency="USD"
                                >
                                  <SickButton>Checkout</SickButton>
                                </StripeCheckout>
                              )}
                            </Mutation>
                          </>
                        ) : ''
                      }
                    </footer>
                  </CartStyles>
                )}
              </Query>
              )}
            </Mutation>
          )
        }}
      </User>
    );

  }
};

export default Cart;