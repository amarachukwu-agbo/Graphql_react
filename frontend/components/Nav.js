import Link from 'next/link';
import { Mutation } from 'react-apollo';
import NavStyles from './styles/NavStyles';
import User from '../components/User';
import SignOut from '../components/SignOut';
import CartCount from '../components/CartCount';
import { TOGGLE_CART_MUTATION } from '../components/Cart';

const Nav = () => (
  <Mutation mutation={TOGGLE_CART_MUTATION}>
    {(toggleCart) => (
      <User>
        {({ data: { me }}) => (
          <NavStyles>
            <Link href="/allitems">
              <a>Shop</a>
            </Link>
            {
              me && (
                <>
                  <Link href="/sell">
                    <a>Sell!</a>
                  </Link>
                  <Link href="/orders">
                    <a>Orders</a>
                  </Link>
                  <Link href="/me">
                    <a>Account</a>
                  </Link>
                  <SignOut />
                  <button onClick={toggleCart}>
                    My Cart
                    <CartCount
                      count={me.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)}
                    />
                  </button>
                </>
              )
            }
            {
              !me && (
                <Link href="/signup">
                  <a>Sign In</a>
                </Link>
              )
            }
      </NavStyles>
        )}
      </User>
    )}
  </Mutation>
);

export default Nav;
