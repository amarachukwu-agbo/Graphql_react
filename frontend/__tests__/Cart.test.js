import { mount } from 'enzyme';
import wait from 'waait';
import NProgress from 'nprogress';
import { MockedProvider } from 'react-apollo/test-utils';
import Cart, { LOCAL_STATE_QUERY, CREATE_ORDER } from '../components/Cart';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { CURRENT_USER_QUERY } from '../components/User';
import Router from 'next/router';

jest.mock('next/router', ()=> ({push: jest.fn()}))


const fakeUserDetails = fakeUser();
const fakeCart = fakeCartItem();
const mocks = [
  {
    request: {
      query: CURRENT_USER_QUERY,
    },
    result: {
      data: {
        me: { ...fakeUserDetails, cart: [fakeCart]  }
      }
    }
  },
  {
    request: {
      query: LOCAL_STATE_QUERY,
    },
    result: {
      data: {
        cartOpen: true
      }
    }
  }
];

describe('<Cart />', () => {
  it('renders correctly and matches snapshot', async () => {
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <Cart />
    </MockedProvider>);
    await wait();
    wrapper.update();
    expect(wrapper.find('header')).toMatchSnapshot();
    expect(wrapper.find('CartItem').length).toBe(1);
    expect(wrapper.find('ReactStripeCheckout')).toMatchSnapshot();
  });

  it('creates an order', async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: {
          id: '234444'
        }
      }
    });
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <Cart />
    </MockedProvider>);
    wrapper.find('Cart').instance().onToken({id: 'abc123'}, createOrderMock);
    expect(createOrderMock).toHaveBeenCalled();
    expect(createOrderMock).toHaveBeenCalledWith({
      variables: {
        token: 'abc123',
      }
    });
  });

  it('starts NProgress when order is created', async () => {
    NProgress.start = jest.fn();
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: {
          id: '234444'
        }
      }
    });
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <Cart />
    </MockedProvider>);
    wrapper.find('Cart').instance().onToken({id: 'abc123'}, createOrderMock);
    expect(NProgress.start).toHaveBeenCalled();
  });

  it('routes to the order page after a component is created', async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: {
          id: '234444'
        }
      }
    });
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <Cart />
    </MockedProvider>);
    wrapper.find('Cart').instance().onToken({id: 'abc123'}, createOrderMock);
    await wait();
    expect(Router.push).toHaveBeenCalled();
  });
});
