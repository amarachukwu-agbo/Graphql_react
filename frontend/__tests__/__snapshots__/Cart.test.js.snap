// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders correctly and matches snapshot 1`] = `
<header>
  <CloseButton
    onClick={[Function]}
    title="close"
  >
    <button
      className="CloseButton-sc-1seb878-0 lbNKfp"
      onClick={[Function]}
      title="close"
    >
      Ã—
    </button>
  </CloseButton>
  <Supreme>
    <h3
      className="Supreme-xv30qb-0 hpaXsq"
    >
      Miss Coleman Berge
      's Cart
    </h3>
  </Supreme>
  <p>
    You have 
    1
     item
     in your cart
  </p>
</header>
`;

exports[`<Cart /> renders correctly and matches snapshot 2`] = `
<ReactStripeCheckout
  ComponentClass="span"
  amount={15000}
  className="StripeCheckout"
  currency="USD"
  description="Order of 3 items"
  image="dog-small.jpg"
  label="Pay With Card"
  locale="auto"
  name="Sick Fits"
  reconfigureOnUpdate={false}
  stripeKey="pk_test_4ApAJihRk5yShkglGcJ0ZSXV00oEwKYLu4"
  token={[Function]}
  triggerEvent="onClick"
>
  <span
    onClick={[Function]}
  >
    <SickButton>
      <button
        className="SickButton-l04z44-0 kouXDz"
      >
        Checkout
      </button>
    </SickButton>
  </span>
</ReactStripeCheckout>
`;
